{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "261b52c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import ehrapy as ep\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "import os\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "32aaa473",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_id_col = \"scRNASeq_sample_ID\"\n",
    "cell_type_key = \"Annotation_major_subset\"\n",
    "samples_metadata_cols = [\"Source\", \"Outcome\", \"Death28\", \"Institute\", \"Pool_ID\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92da8a22",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA_PATH = \"/Users/farhad/helmholtz/patpy/pat_rep_benchmark/data/combat_processed_200.h5ad\"\n",
    "HVG_CSV = \"hvg_genes.csv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "58180106",
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE_ADATAS_PATH = \"sample_adatas_500gene\"\n",
    "os.makedirs(SAMPLE_ADATAS_PATH, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "16e791ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "dirs = [\n",
    "    \"logs/pretrain\",\n",
    "    \"logs/finetune\",\n",
    "    \"ckpts/pretrain\",\n",
    "    \"ckpts/finetune\"\n",
    "]\n",
    "\n",
    "# Create each directory\n",
    "for d in dirs:\n",
    "    os.makedirs(d, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c193103e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 27588 × 3000\n",
       "    obs: 'Annotation_cluster_id', 'Annotation_cluster_name', 'Annotation_minor_subset', 'Annotation_major_subset', 'Annotation_cell_type', 'GEX_region', 'QC_ngenes', 'QC_total_UMI', 'QC_pct_mitochondrial', 'QC_scrub_doublet_scores', 'TCR_chain_composition', 'TCR_clone_ID', 'TCR_clone_count', 'TCR_clone_proportion', 'TCR_contains_unproductive', 'TCR_doublet', 'TCR_chain_TRA', 'TCR_v_gene_TRA', 'TCR_d_gene_TRA', 'TCR_j_gene_TRA', 'TCR_c_gene_TRA', 'TCR_productive_TRA', 'TCR_cdr3_TRA', 'TCR_umis_TRA', 'TCR_chain_TRA2', 'TCR_v_gene_TRA2', 'TCR_d_gene_TRA2', 'TCR_j_gene_TRA2', 'TCR_c_gene_TRA2', 'TCR_productive_TRA2', 'TCR_cdr3_TRA2', 'TCR_umis_TRA2', 'TCR_chain_TRB', 'TCR_v_gene_TRB', 'TCR_d_gene_TRB', 'TCR_j_gene_TRB', 'TCR_c_gene_TRB', 'TCR_productive_TRB', 'TCR_chain_TRB2', 'TCR_v_gene_TRB2', 'TCR_d_gene_TRB2', 'TCR_j_gene_TRB2', 'TCR_c_gene_TRB2', 'TCR_productive_TRB2', 'TCR_cdr3_TRB2', 'TCR_umis_TRB2', 'BCR_umis_HC', 'BCR_contig_qc_HC', 'BCR_functionality_HC', 'BCR_v_call_HC', 'BCR_v_score_HC', 'BCR_j_call_HC', 'BCR_j_score_HC', 'BCR_junction_aa_HC', 'BCR_total_mut_HC', 'BCR_s_mut_HC', 'BCR_r_mut_HC', 'BCR_c_gene_HC', 'BCR_clone_per_replicate_HC', 'BCR_clone_global_HC', 'BCR_clonal_abundance_HC', 'BCR_locus_LC', 'BCR_umis_LC', 'BCR_contig_qc_LC', 'BCR_functionality_LC', 'BCR_v_call_LC', 'BCR_v_score_LC', 'BCR_j_call_LC', 'BCR_j_score_LC', 'BCR_junction_aa_LC', 'BCR_total_mut_LC', 'BCR_s_mut_LC', 'BCR_r_mut_LC', 'BCR_c_gene_LC', 'COMBAT_ID', 'scRNASeq_sample_ID', 'COMBAT_participant_timepoint_ID', 'Source', 'Age', 'Sex', 'Race', 'BMI', 'Hospitalstay', 'Death28', 'Institute', 'PreExistingHeartDisease', 'PreExistingLungDisease', 'PreExistingKidneyDisease', 'PreExistingDiabetes', 'PreExistingHypertension', 'PreExistingImmunocompromised', 'Smoking', 'Symptomatic', 'Requiredvasoactive', 'Respiratorysupport', 'SARSCoV2PCR', 'Outcome', 'TimeSinceOnset', 'Ethnicity', 'Tissue', 'DiseaseClassification', 'Pool_ID', 'Channel_ID', 'ifn_1_score', '_scvi_batch', '_scvi_labels', 'dpt_pseudotime'\n",
       "    var: 'gene_ids', 'feature_types', 'highly_variable', 'highly_variable_rank', 'means', 'variances', 'variances_norm', 'highly_variable_nbatches'\n",
       "    uns: 'Institute', 'ObjectCreateDate', 'Source_colors', 'Technology', '_scvi_manager_uuid', '_scvi_uuid', 'diffmap_evals', 'genome_annotation_version', 'hvg', 'iroot', 'log1p', 'neighbors', 'pca'\n",
       "    obsm: 'X_diffmap', 'X_harmony', 'X_pca', 'X_raw_counts', 'X_scANVI_Pool_ID', 'X_scANVI_scRNASeq_sample_ID', 'X_scVI_Pool_ID', 'X_scVI_scRNASeq_sample_ID', 'X_umap', 'X_umap_source'\n",
       "    varm: 'PCs'\n",
       "    layers: 'X_raw_counts'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = sc.read_h5ad(DATA_PATH)\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c3681543",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 1])"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.Death28.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c15729be",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.X = adata.layers[\"X_raw_counts\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a41b8b53",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/farhad/miniconda3/envs/mcBERT/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:75: UserWarning: `flavor='seurat_v3'` expects raw count data, but non-integers were found.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "hvg_info = sc.pp.highly_variable_genes(\n",
    "    adata, flavor=\"seurat_v3\", n_top_genes=500, inplace=False\n",
    ")\n",
    "\n",
    "hvg_genes = adata.var_names[hvg_info.highly_variable]\n",
    "\n",
    "# Save to CSV\n",
    "pd.DataFrame({\"genes\": hvg_genes}).to_csv(\"hvg_genes.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d9ea46ae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing: S00109-Ja001E-PBCa\n",
      "Saved to sample_adatas_500gene/S00109-Ja001E-PBCa.h5ad\n",
      "Processing: S00112-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00112-Ja003E-PBCa.h5ad\n",
      "Processing: S00005-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/S00005-Ja005E-PBCa.h5ad\n",
      "Processing: S00061-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00061-Ja003E-PBCa.h5ad\n",
      "Processing: S00056-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00056-Ja003E-PBCa.h5ad\n",
      "Processing: N00027-Ja001E-PBGa\n",
      "Saved to sample_adatas_500gene/N00027-Ja001E-PBGa.h5ad\n",
      "Processing: H00067-Ha001E-PBGa\n",
      "Saved to sample_adatas_500gene/H00067-Ha001E-PBGa.h5ad\n",
      "Processing: G05153-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/G05153-Ja005E-PBCa.h5ad\n",
      "Processing: U00515-Ua005E-PBUa\n",
      "Saved to sample_adatas_500gene/U00515-Ua005E-PBUa.h5ad\n",
      "Processing: U00505-Ua005E-PBUa\n",
      "Saved to sample_adatas_500gene/U00505-Ua005E-PBUa.h5ad\n",
      "Processing: S00028-Ja001E-PBCa\n",
      "Saved to sample_adatas_500gene/S00028-Ja001E-PBCa.h5ad\n",
      "Processing: N00023-Ja001E-PBGa\n",
      "Saved to sample_adatas_500gene/N00023-Ja001E-PBGa.h5ad\n",
      "Processing: G05064-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/G05064-Ja005E-PBCa.h5ad\n",
      "Processing: S00043-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/S00043-Ja005E-PBCa.h5ad\n",
      "Processing: S00027-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00027-Ja003E-PBCa.h5ad\n",
      "Processing: G05097-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/G05097-Ja005E-PBCa.h5ad\n",
      "Processing: S00033-Ja001E-PBCa\n",
      "Saved to sample_adatas_500gene/S00033-Ja001E-PBCa.h5ad\n",
      "Processing: N00018-Ja003E-PBGa\n",
      "Saved to sample_adatas_500gene/N00018-Ja003E-PBGa.h5ad\n",
      "Processing: S00040-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/S00040-Ja005E-PBCa.h5ad\n",
      "Processing: H00054-Ha001E-PBGa\n",
      "Saved to sample_adatas_500gene/H00054-Ha001E-PBGa.h5ad\n",
      "Processing: S00034-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/S00034-Ja005E-PBCa.h5ad\n",
      "Processing: G05164-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/G05164-Ja005E-PBCa.h5ad\n",
      "Processing: S00078-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00078-Ja003E-PBCa.h5ad\n",
      "Processing: S00104-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00104-Ja003E-PBCa.h5ad\n",
      "Processing: S00119-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00119-Ja003E-PBCa.h5ad\n",
      "Processing: U00519-Ua005E-PBUa\n",
      "Saved to sample_adatas_500gene/U00519-Ua005E-PBUa.h5ad\n",
      "Processing: N00024-Ja001E-PBGa\n",
      "Saved to sample_adatas_500gene/N00024-Ja001E-PBGa.h5ad\n",
      "Processing: U00613-Ua005E-PBUa\n",
      "Saved to sample_adatas_500gene/U00613-Ua005E-PBUa.h5ad\n",
      "Processing: S00016-Ja001E-PBCa\n",
      "Saved to sample_adatas_500gene/S00016-Ja001E-PBCa.h5ad\n",
      "Processing: G05061-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/G05061-Ja005E-PBCa.h5ad\n",
      "Processing: S00039-Ja001E-PBCa\n",
      "Saved to sample_adatas_500gene/S00039-Ja001E-PBCa.h5ad\n",
      "Processing: N00032-Ja001E-PBGa\n",
      "Saved to sample_adatas_500gene/N00032-Ja001E-PBGa.h5ad\n",
      "Processing: S00052-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/S00052-Ja005E-PBCa.h5ad\n",
      "Processing: N00028-Ja001E-PBGa\n",
      "Saved to sample_adatas_500gene/N00028-Ja001E-PBGa.h5ad\n",
      "Processing: S00129-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/S00129-Ja005E-PBCa.h5ad\n",
      "Processing: H00085-Ha001E-PBGa\n",
      "Saved to sample_adatas_500gene/H00085-Ha001E-PBGa.h5ad\n",
      "Processing: S00042-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00042-Ja003E-PBCa.h5ad\n",
      "Processing: S00058-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00058-Ja003E-PBCa.h5ad\n",
      "Processing: S00020-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00020-Ja003E-PBCa.h5ad\n",
      "Processing: U00607-Ua005E-PBUa\n",
      "Saved to sample_adatas_500gene/U00607-Ua005E-PBUa.h5ad\n",
      "Processing: S00056-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/S00056-Ja005E-PBCa.h5ad\n",
      "Processing: G05092-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/G05092-Ja005E-PBCa.h5ad\n",
      "Processing: N00033-Ja001E-PBGa\n",
      "Saved to sample_adatas_500gene/N00033-Ja001E-PBGa.h5ad\n",
      "Processing: S00006-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/S00006-Ja005E-PBCa.h5ad\n",
      "Processing: G05078-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/G05078-Ja005E-PBCa.h5ad\n",
      "Processing: S00109-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00109-Ja003E-PBCa.h5ad\n",
      "Processing: H00049-Ha001E-PBGa\n",
      "Saved to sample_adatas_500gene/H00049-Ha001E-PBGa.h5ad\n",
      "Processing: N00029-Ja005E-PBGa\n",
      "Saved to sample_adatas_500gene/N00029-Ja005E-PBGa.h5ad\n",
      "Processing: S00041-Ja001E-PBCa\n",
      "Saved to sample_adatas_500gene/S00041-Ja001E-PBCa.h5ad\n",
      "Processing: S00045-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00045-Ja003E-PBCa.h5ad\n",
      "Processing: S00005-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00005-Ja003E-PBCa.h5ad\n",
      "Processing: S00068-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/S00068-Ja005E-PBCa.h5ad\n",
      "Processing: N00021-Ja001E-PBGa\n",
      "Saved to sample_adatas_500gene/N00021-Ja001E-PBGa.h5ad\n",
      "Processing: U00701-Ua005E-PBUa\n",
      "Saved to sample_adatas_500gene/U00701-Ua005E-PBUa.h5ad\n",
      "Processing: S00007-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00007-Ja003E-PBCa.h5ad\n",
      "Processing: S00048-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/S00048-Ja005E-PBCa.h5ad\n",
      "Processing: S00003-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00003-Ja003E-PBCa.h5ad\n",
      "Processing: N00039-Ja001E-PBGa\n",
      "Saved to sample_adatas_500gene/N00039-Ja001E-PBGa.h5ad\n",
      "Processing: S00097-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00097-Ja003E-PBCa.h5ad\n",
      "Processing: G05171-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/G05171-Ja005E-PBCa.h5ad\n",
      "Processing: H00058-Ha001E-PBGa\n",
      "Saved to sample_adatas_500gene/H00058-Ha001E-PBGa.h5ad\n",
      "Processing: S00037-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00037-Ja003E-PBCa.h5ad\n",
      "Processing: N00040-Ja001E-PBGa\n",
      "Saved to sample_adatas_500gene/N00040-Ja001E-PBGa.h5ad\n",
      "Processing: S00064-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/S00064-Ja005E-PBCa.h5ad\n",
      "Processing: S00042-Ja001E-PBCa\n",
      "Saved to sample_adatas_500gene/S00042-Ja001E-PBCa.h5ad\n",
      "Processing: U00605-Ua005E-PBUa\n",
      "Saved to sample_adatas_500gene/U00605-Ua005E-PBUa.h5ad\n",
      "Processing: N00047-Ja005E-PBGa\n",
      "Saved to sample_adatas_500gene/N00047-Ja005E-PBGa.h5ad\n",
      "Processing: S00033-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/S00033-Ja005E-PBCa.h5ad\n",
      "Processing: S00095-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/S00095-Ja005E-PBCa.h5ad\n",
      "Processing: S00008-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00008-Ja003E-PBCa.h5ad\n",
      "Processing: S00045-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/S00045-Ja005E-PBCa.h5ad\n",
      "Processing: G05077-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/G05077-Ja005E-PBCa.h5ad\n",
      "Processing: S00134-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00134-Ja003E-PBCa.h5ad\n",
      "Processing: S00050-Ja001E-PBCa\n",
      "Saved to sample_adatas_500gene/S00050-Ja001E-PBCa.h5ad\n",
      "Processing: H00072-Ha001E-PBGa\n",
      "Saved to sample_adatas_500gene/H00072-Ha001E-PBGa.h5ad\n",
      "Processing: S00094-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/S00094-Ja005E-PBCa.h5ad\n",
      "Processing: N00049-Ja005E-PBGa\n",
      "Saved to sample_adatas_500gene/N00049-Ja005E-PBGa.h5ad\n",
      "Processing: S00113-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/S00113-Ja005E-PBCa.h5ad\n",
      "Processing: N00006-Ja003E-PBGa\n",
      "Saved to sample_adatas_500gene/N00006-Ja003E-PBGa.h5ad\n",
      "Processing: U00619-Ua005E-PBUa\n",
      "Saved to sample_adatas_500gene/U00619-Ua005E-PBUa.h5ad\n",
      "Processing: S00148-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00148-Ja003E-PBCa.h5ad\n",
      "Processing: S00129-Ja001E-PBCa\n",
      "Saved to sample_adatas_500gene/S00129-Ja001E-PBCa.h5ad\n",
      "Processing: S00060-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/S00060-Ja005E-PBCa.h5ad\n",
      "Processing: N00012-Ja003E-PBGa\n",
      "Saved to sample_adatas_500gene/N00012-Ja003E-PBGa.h5ad\n",
      "Processing: S00063-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00063-Ja003E-PBCa.h5ad\n",
      "Processing: H00052-Ha001E-PBGa\n",
      "Saved to sample_adatas_500gene/H00052-Ha001E-PBGa.h5ad\n",
      "Processing: S00106-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00106-Ja003E-PBCa.h5ad\n",
      "Processing: N00025-Ja001E-PBGa\n",
      "Saved to sample_adatas_500gene/N00025-Ja001E-PBGa.h5ad\n",
      "Processing: S00067-Ja001E-PBCa\n",
      "Saved to sample_adatas_500gene/S00067-Ja001E-PBCa.h5ad\n",
      "Processing: N00007-Ja003E-PBGa\n",
      "Saved to sample_adatas_500gene/N00007-Ja003E-PBGa.h5ad\n",
      "Processing: S00049-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00049-Ja003E-PBCa.h5ad\n",
      "Processing: G05145-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/G05145-Ja005E-PBCa.h5ad\n",
      "Processing: S00109-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/S00109-Ja005E-PBCa.h5ad\n",
      "Processing: S00081-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/S00081-Ja005E-PBCa.h5ad\n",
      "Processing: S00024-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00024-Ja003E-PBCa.h5ad\n",
      "Processing: S00077-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/S00077-Ja005E-PBCa.h5ad\n",
      "Processing: N00050-Ja001E-PBGa\n",
      "Saved to sample_adatas_500gene/N00050-Ja001E-PBGa.h5ad\n",
      "Processing: U00503-Ua005E-PBUa\n",
      "Saved to sample_adatas_500gene/U00503-Ua005E-PBUa.h5ad\n",
      "Processing: S00114-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00114-Ja003E-PBCa.h5ad\n",
      "Processing: N00028-Ja005E-PBGa\n",
      "Saved to sample_adatas_500gene/N00028-Ja005E-PBGa.h5ad\n",
      "Processing: H00070-Ha001E-PBGa\n",
      "Saved to sample_adatas_500gene/H00070-Ha001E-PBGa.h5ad\n",
      "Processing: S00052-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00052-Ja003E-PBCa.h5ad\n",
      "Processing: S00099-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/S00099-Ja005E-PBCa.h5ad\n",
      "Processing: N00023-Ja003E-PBGa\n",
      "Saved to sample_adatas_500gene/N00023-Ja003E-PBGa.h5ad\n",
      "Processing: S00126-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00126-Ja003E-PBCa.h5ad\n",
      "Processing: S00142-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/S00142-Ja005E-PBCa.h5ad\n",
      "Processing: G05105-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/G05105-Ja005E-PBCa.h5ad\n",
      "Processing: S00069-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/S00069-Ja005E-PBCa.h5ad\n",
      "Processing: N00017-Ja003E-PBGa\n",
      "Saved to sample_adatas_500gene/N00017-Ja003E-PBGa.h5ad\n",
      "Processing: S00054-Ja001E-PBCa\n",
      "Saved to sample_adatas_500gene/S00054-Ja001E-PBCa.h5ad\n",
      "Processing: U00617-Ua005E-PBUa\n",
      "Saved to sample_adatas_500gene/U00617-Ua005E-PBUa.h5ad\n",
      "Processing: S00124-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00124-Ja003E-PBCa.h5ad\n",
      "Processing: S00081-Ja001E-PBCa\n",
      "Saved to sample_adatas_500gene/S00081-Ja001E-PBCa.h5ad\n",
      "Processing: S00059-Ja001E-PBCa\n",
      "Saved to sample_adatas_500gene/S00059-Ja001E-PBCa.h5ad\n",
      "Processing: G05073-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/G05073-Ja005E-PBCa.h5ad\n",
      "Processing: H00053-Ha001E-PBGa\n",
      "Saved to sample_adatas_500gene/H00053-Ha001E-PBGa.h5ad\n",
      "Processing: S00033-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00033-Ja003E-PBCa.h5ad\n",
      "Processing: S00124-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/S00124-Ja005E-PBCa.h5ad\n",
      "Processing: N00013-Ja005E-PBGa\n",
      "Saved to sample_adatas_500gene/N00013-Ja005E-PBGa.h5ad\n",
      "Processing: S00066-Ja001E-PBCa\n",
      "Saved to sample_adatas_500gene/S00066-Ja001E-PBCa.h5ad\n",
      "Processing: S00096-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/S00096-Ja005E-PBCa.h5ad\n",
      "Processing: S00053-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00053-Ja003E-PBCa.h5ad\n",
      "Processing: N00044-Ja001E-PBGa\n",
      "Saved to sample_adatas_500gene/N00044-Ja001E-PBGa.h5ad\n",
      "Processing: U00601-Ua005E-PBUa\n",
      "Saved to sample_adatas_500gene/U00601-Ua005E-PBUa.h5ad\n",
      "Processing: U00504-Ua005E-PBUa\n",
      "Saved to sample_adatas_500gene/U00504-Ua005E-PBUa.h5ad\n",
      "Processing: S00030-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00030-Ja003E-PBCa.h5ad\n",
      "Processing: S00082-Ja001E-PBCa\n",
      "Saved to sample_adatas_500gene/S00082-Ja001E-PBCa.h5ad\n",
      "Processing: S00057-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00057-Ja003E-PBCa.h5ad\n",
      "Processing: H00064-Ha001E-PBGa\n",
      "Saved to sample_adatas_500gene/H00064-Ha001E-PBGa.h5ad\n",
      "Processing: S00045-Ja001E-PBCa\n",
      "Saved to sample_adatas_500gene/S00045-Ja001E-PBCa.h5ad\n",
      "Processing: N00037-Ja003E-PBGa\n",
      "Saved to sample_adatas_500gene/N00037-Ja003E-PBGa.h5ad\n",
      "Processing: S00002-Ja001E-PBCa\n",
      "Saved to sample_adatas_500gene/S00002-Ja001E-PBCa.h5ad\n",
      "Processing: S00076-Ja001E-PBCa\n",
      "Saved to sample_adatas_500gene/S00076-Ja001E-PBCa.h5ad\n",
      "Processing: S00072-Ja001E-PBCa\n",
      "Saved to sample_adatas_500gene/S00072-Ja001E-PBCa.h5ad\n",
      "Processing: U00502-Ua005E-PBUa\n",
      "Saved to sample_adatas_500gene/U00502-Ua005E-PBUa.h5ad\n",
      "Processing: S00065-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00065-Ja003E-PBCa.h5ad\n",
      "Processing: S00048-Ja003E-PBCa\n",
      "Saved to sample_adatas_500gene/S00048-Ja003E-PBCa.h5ad\n",
      "Processing: G05112-Ja005E-PBCa\n",
      "Saved to sample_adatas_500gene/G05112-Ja005E-PBCa.h5ad\n",
      "Processing: N00038-Ja001E-PBGa\n",
      "Saved to sample_adatas_500gene/N00038-Ja001E-PBGa.h5ad\n",
      "Processing: U00501-Ua005E-PBUa\n",
      "Saved to sample_adatas_500gene/U00501-Ua005E-PBUa.h5ad\n"
     ]
    }
   ],
   "source": [
    "SELECTED_GENES = np.array(pd.read_csv(\"hvg_genes.csv\")[\"genes\"])\n",
    "DONOR_COLUMN = sample_id_col\n",
    "\n",
    "def process_donor(adata,donor):\n",
    "    print(f\"Processing: {donor}\")\n",
    "    donor_h5 = adata[adata.obs[DONOR_COLUMN] == donor]\n",
    "    save_path = SAMPLE_ADATAS_PATH + f\"/{donor}.h5ad\"\n",
    "    sc.write(save_path, donor_h5)\n",
    "    print(f\"Saved to {save_path}\")\n",
    "\n",
    "\n",
    "\n",
    "sc.pp.normalize_total(adata, target_sum=1, inplace=True)\n",
    "donors = adata.obs[DONOR_COLUMN].unique()\n",
    "\n",
    "for donor in donors:\n",
    "    process_donor(adata, donor)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a5a073f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Epoch: 1/2        0%|                                                           RobertaSdpaSelfAttention is used but `torch.nn.functional.scaled_dot_product_attention` does not support non-absolute `position_embedding_type` or `output_attentions=True` or `head_mask`. Falling back to the manual attention implementation, but specifying the manual implementation will be required from Transformers version v5.0.0 onwards. This warning can be removed using the argument `attn_implementation=\"eager\"` when loading the model.\n",
      "Epoch: 1/2      100%|###########################################################\n",
      "Evaluating...   100%|###########################################################\n",
      "\n",
      "Epoch: 2/2      100%|###########################################################\n",
      "Evaluating...   100%|###########################################################\n",
      "Saved checkpoint to `ckpts/pretrain/2.pt`\n"
     ]
    }
   ],
   "source": [
    "!python pretrain.py --config configs/configs_pretraining/Data2Vec_heart_example.yaml\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "c5b19d54",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using 98 patients for training and 14 patients for validation representing 2 unique disease\n",
      "Training diseases:  [0 1]\n",
      "Epoch 0, loss: 0.0000:   0%|                              | 0/4 [00:19<?, ?it/s]RobertaSdpaSelfAttention is used but `torch.nn.functional.scaled_dot_product_attention` does not support non-absolute `position_embedding_type` or `output_attentions=True` or `head_mask`. Falling back to the manual attention implementation, but specifying the manual implementation will be required from Transformers version v5.0.0 onwards. This warning can be removed using the argument `attn_implementation=\"eager\"` when loading the model.\n",
      "Epoch 0, loss: 2.8876: 100%|█████████████████████| 4/4 [23:19<00:00, 349.98s/it]\n",
      "100%|█████████████████████████████████████████████| 4/4 [03:57<00:00, 59.44s/it]\n",
      "  0%|                                                     | 0/1 [00:00<?, ?it/s]python(8445) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.\n",
      "python(8446) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.\n",
      "python(8447) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.\n",
      "python(8448) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.\n",
      "Epoch 0, val_loss: 0.0000: 100%|██████████████████| 1/1 [00:37<00:00, 37.58s/it]\n",
      "Epoch 1, loss: 2.8876: 100%|█████████████████████| 4/4 [22:25<00:00, 336.35s/it]\n",
      "100%|█████████████████████████████████████████████| 4/4 [03:48<00:00, 57.23s/it]\n",
      "Epoch 1, val_loss: 0.0000: 100%|██████████████████| 1/1 [00:21<00:00, 21.06s/it]\n"
     ]
    }
   ],
   "source": [
    "!python train.py --config configs/configs_finetuning/FineTune_heart_example.yaml\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "0954b2d5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "python(19439) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using 140 patients representing [0 1] diseases\n",
      "  0%|                                                    | 0/18 [00:00<?, ?it/s]RobertaSdpaSelfAttention is used but `torch.nn.functional.scaled_dot_product_attention` does not support non-absolute `position_embedding_type` or `output_attentions=True` or `head_mask`. Falling back to the manual attention implementation, but specifying the manual implementation will be required from Transformers version v5.0.0 onwards. This warning can be removed using the argument `attn_implementation=\"eager\"` when loading the model.\n",
      "100%|███████████████████████████████████████████| 18/18 [01:56<00:00,  6.49s/it]\n",
      "/Users/farhad/miniconda3/envs/mcBERT/lib/python3.9/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.\n",
      "  warn(\n"
     ]
    }
   ],
   "source": [
    "!python inference.py --config configs/configs_inference/Inference_heart_example.yaml\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mcBERT",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
